File: backend/chat-application/src/main/java/com/synapse/chat_application/controller/ChannelController.java
package com.synapse.chat_application.controller;

import com.synapse.chat_application.dto.guild.ChannelRequest;
import com.synapse.chat_application.dto.guild.ChannelResponse;
import com.synapse.chat_application.service.ChannelService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.UUID;

@RestController
@RequestMapping("/api/guilds/{guildId}/channels")
@RequiredArgsConstructor
public class ChannelController {

    private final ChannelService channelService;

    @PostMapping
    public ResponseEntity<ChannelResponse> createChannel(
            @AuthenticationPrincipal UserDetails userDetails,
            @PathVariable UUID guildId,
            @Valid @RequestBody ChannelRequest request) {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(channelService.createChannel(userDetails.getUsername(), guildId, request));
    }

    @GetMapping
    public ResponseEntity<List<ChannelResponse>> getChannels(
            @AuthenticationPrincipal UserDetails userDetails,
            @PathVariable UUID guildId) {
        return ResponseEntity.ok(channelService.getChannels(guildId, userDetails.getUsername()));
    }

    @DeleteMapping("/{channelId}")
    public ResponseEntity<Void> deleteChannel(
            @AuthenticationPrincipal UserDetails userDetails,
            @PathVariable UUID guildId,
            @PathVariable UUID channelId) {
        channelService.deleteChannel(userDetails.getUsername(), guildId, channelId);
        return ResponseEntity.noContent().build();
    }
}

File: backend/backend/src/controllers/channel.controller.js
import channelService from "../services/channel.service.js";
import { asyncHandler } from "../utils/asyncHandler.js";
import { ApiResponse } from "../utils/ApiResponse.js";
import { createChannelSchema } from "../validators/guild.validators.js";

export const createChannel = asyncHandler(async (req, res) => {
    const validatedData = createChannelSchema.parse(req.body);
    const channel = await channelService.createChannel(
        req.params.guildId, 
        validatedData
    );
    
    return res.status(201).json(
        new ApiResponse(201, channel, "Channel created successfully")
    );
});

export const getChannels = asyncHandler(async (req, res) => {
    const channels = await channelService.getChannels(req.params.guildId);
    
    return res.status(200).json(
        new ApiResponse(200, channels, "Channels fetched successfully")
    );
});

export const deleteChannel = asyncHandler(async (req, res) => {
    await channelService.deleteChannel(req.params.channelId);
    
    return res.status(200).json(
        new ApiResponse(200, {}, "Channel deleted successfully")
    );
});
